### PR TITLE
CP-47012: change pending guidance in old xapi to recommended ones in new xapi

### PR DESCRIPTION
**Pull request by gangj**
Add update guidance DB upgrade rule: except reboot_host_on_livepatch_failure, move any guidance in
"host.pending_guidances" into "host.pending_guidances_recommended".

### CODE REVIEW

File: ocaml/xapi/xapi_db_upgrade.ml [hunk 1]
```diff
@@ -862,10 +862,11 @@ let empty_pool_uefi_certificates =
...
     description=
-      "Replace reboot_host_on_livepatch_failure in host.pending_guidances with \
-       reboot_host_on_kernel_livepatch_failure and \
+      "1. Replace reboot_host_on_livepatch_failure in host.pending_guidances \
+       with reboot_host_on_kernel_livepatch_failure and \
        reboot_host_on_xen_livepatch_failure in \
-       host.pending_guidances_recommended"
+       host.pending_guidances_recommended. 2. Move the rest guidances in \
+       host.pending_guidances into host.pending_guidances_recommended"
   ; version= (fun _ -> true)
```
Comment (robhoes) [inline ocaml/xapi/xapi_db_upgrade.ml:870]: Careful here: we don't want to do this for _every_ DB upgrade, only when moving from the old schema to the new.

Comment (gangj) [reply ocaml/xapi/xapi_db_upgrade.ml]: Sorry, forgot that!

File: ocaml/xapi/xapi_db_upgrade.ml [hunk 2]
```diff
@@ -862,11 +862,13 @@ let empty_pool_uefi_certificates =
 let update_livepatch_guidance =
```
Comment (minglumlu) [inline ocaml/xapi/xapi_db_upgrade.ml:862]: You may need to change the name of the function.

File: ocaml/xapi/xapi_db_upgrade.ml [hunk 3]
```diff
@@ -881,7 +883,14 @@ let update_livepatch_guidance =
                    ~value:`reboot_host_on_xen_livepatch_failure ;
                  Db.Host.remove_pending_guidances ~__context ~self
```
Comment (minglumlu) [inline ocaml/xapi/xapi_db_upgrade.ml:884]: This is no longer necessary as the list will be reset to empty anyway.

File: ocaml/xapi/xapi_db_upgrade.ml [hunk 4]
```diff
@@ -862,11 +862,13 @@ let empty_pool_uefi_certificates =
 let update_livepatch_guidance =
   {
     description=
-      "Replace reboot_host_on_livepatch_failure in host.pending_guidances with \
-       reboot_host_on_kernel_livepatch_failure and \
+      "1. Replace reboot_host_on_livepatch_failure in host.pending_guidances \
```
Comment (minglumlu) [inline ocaml/xapi/xapi_db_upgrade.ml:865]: This description seems too long to be in log. Could it be a one-line sentence and change this paragraph to comment?

**Review by minglumlu — APPROVED** [review]

File: ocaml/xapi/xapi_db_upgrade.ml [hunk 5]
```diff
@@ -881,7 +885,14 @@ let update_livepatch_guidance =
...
                    ~value:`reboot_host_on_livepatch_failure
-               )
+               ) ;
+               List.iter
+                 (fun g ->
+                   Db.Host.add_pending_guidances_recommended ~__context ~self
+                     ~value:g
+                 )
+                 (Db.Host.get_pending_guidances ~__context ~self) ;
+               Db.Host.set_pending_guidances ~__context ~self ~value:[]
```
Comment (minglumlu) [inline ocaml/xapi/xapi_db_upgrade.ml:895]: Need to handle the VM.pending_guidances as well.

File: ocaml/xapi/xapi_db_upgrade.ml [hunk 6]
```diff
@@ -867,7 +867,8 @@ let update_livepatch_guidance =
        reboot_host_on_xen_livepatch_failure in \
        host.pending_guidances_recommended. 2. Move the rest guidances in \
        host.pending_guidances into host.pending_guidances_recommended"
-  ; version= (fun _ -> true)
+      (* TODO: update below schema version to which the feature branch got merged with *)
+  ; version= (fun x -> x < (5, 769))
```
Comment (psafont) [inline ocaml/xapi/xapi_db_upgrade.ml:873]: I recommend adding a failure to the CI so this note is not forgotten when merging to master:
```
let update_livepatch_guidance =
  let change_version_to_the_latest_schema_version = () in
  {
    description=
      "1. Replace reboot_host_on_livepatch_failure ....
```

This will mean that there might be some errors creeping before merging to master, but those can be fixed at that time and they tend to be rare

**Comment by gangj** [comment]
I think this might requires more thinking, we can't do this mapping simply this way, as the pending guidances in `host.pending_guidances` from old xapi would not only includes the guidance from old updates, but it might also includes the guidance from new updates!

Let's see this scenario:
1. User check updates with old xapi which has outstanding guidances in `host.pending_guidances`, and finds available updates including "new xapi", and some newer updates after the one for "new xapi", 
2. User choose to apply the updates but not auto-apply recommended guidance(if there are many "recommended guidance" from available updates, it is less likely that user would choose to auto apply the recommended guidances), with XAPI call `host.apply_updates` to old xapi, the rpm packages for all the updates would be installed, and the host guidances of the installed updates would be added into `host.pending_guidances` of old xapi!
3. And the new added DB upgrade rule for update guidance would not be triggered until old xapi is restarted later, when new xapi is started, the guidances in `host.pending_guidances` from old xapi actually include both the "outstanding guidance before install updates" and the guidances from those recently installed updates!

Several ways we might take:
1. for safety, do not move any outstanding guidance from `host.pending_guidances` into `host.pending_guidances_recommended`, but new XC will have user apply all those guidances before doing anything else! This might be a bad experience for our users.
2. keep some known possible mandatory guidances in `host.pending_guidances` (like "restart_toolstack", "restart_device_model"), move others into `host.pending_guidances_recommended`. But we might be hard to determine the possible mandatory guidances for future updates now.

Is this a problem? Is there a better solution?

**Comment by gangj** [comment]
> I think this might requires more thinking, we can't do this mapping simply this way, as the pending guidances in `host.pending_guidances` from old xapi would not only includes the guidance from old updates, but it might also includes the guidance from new updates!
> 
> Let's see this scenario:
> 
> 1. User check updates with old xapi which has outstanding guidances in `host.pending_guidances`, and finds available updates including "new xapi", and some newer updates after the one for "new xapi",
> 2. User choose to apply the updates but not auto-apply recommended guidance(if there are many "recommended guidance" from available updates, it is less likely that user would choose to auto apply the recommended guidances), with XAPI call `host.apply_updates` to old xapi, the rpm packages for all the updates would be installed, and the host guidances of the installed updates would be added into `host.pending_guidances` of old xapi!
> 3. And the new added DB upgrade rule for update guidance would not be triggered until old xapi is restarted later, when new xapi is started, the guidances in `host.pending_guidances` from old xapi actually include both the "outstanding guidance before install updates" and the guidances from those recently installed updates!
> 
> Several ways we might take:
> 
> 1. for safety, do not move any outstanding guidance from `host.pending_guidances` into `host.pending_guidances_recommended`, but new XC will have user apply all those guidances before doing anything else! This might be a bad experience for our users.
> 2. keep some known possible mandatory guidances in `host.pending_guidances` (like "restart_toolstack", "restart_device_model"), move others into `host.pending_guidances_recommended`. But we might be hard to determine the possible mandatory guidances for future updates now.
> 
> Is this a problem? Is there a better solution?

Thank you @minglumlu for the discussion! It should not be a real problem, because:
Currently there are only 2 kinds of host guidance besides livepatch failure guidance: restart_toolstack and reboot_host, for reboot_host, as discussed with factory team, it had not been used as mandatory guidance for old updates, and it will not be used as mandatory guidance for future updates, we can safely move it from "host.pending_guidances" in old xapi to "host.pending_guidances_recommended" in new xapi. While for restart_toolstack, it should have been applied and cleared as new xapi would not take effect without a toolstack restart.

In summary, the added code in this PR actually would only move guidance "reboot_host" from "host.pending_guidances" in old xapi to "host.pending_guidances_recommended" in new xapi, this is expected.

**Comment by gangj** [comment]
> Need to handle the VM.pending_guidances as well. Or it could be ignored as the VM.pending_guidances won't be considered in safety checks?

For VM guidance, we only might have `restart_device_model` in `vm.pending_guidances` in old xapi(the other one `restart_vm` is new added, and it will not be used as mandatory one as discussed with factory team), as `restart_device_model` might be a mandatory one, I think it is safe to keep it as mandatory and not moved.

**Comment by gangj** [comment]
> I recommend adding a failure to the CI so this note is not forgotten when merging to master:
> 
> ```
> let update_livepatch_guidance =
>   let change_version_to_the_latest_schema_version = () in
>   {
>     description=
>       "1. Replace reboot_host_on_livepatch_failure ....
> ```
> 
> This will mean that there might be some errors creeping before merging to master, but those can be fixed at that time and they tend to be rare

@psafont , sorry, adding such a failure will also make the build here fail, do you mean we can merge this PR into feature branch ignoring this build failure here?

**Comment by psafont** [comment]
> @psafont , sorry, adding such a failure will also make the build here fail, do you mean we can merge this PR into feature branch ignoring this build failure here?

yes

**Comment by gangj** [comment]
To reduce the build failure impact on following PRs, @minglumlu suggest to merge this PR as late as possible.

**Comment by gangj** [comment]
Force push to rebase the commits in this PR to the latest feature branch.

File: ocaml/xapi/xapi_db_upgrade.ml [hunk 7]
```diff
@@ -859,14 +859,20 @@ let empty_pool_uefi_certificates =
...
+let upgrade_update_guidance =
   {
     description=
-      "Replace reboot_host_on_livepatch_failure in host.pending_guidances with \
-       reboot_host_on_kernel_livepatch_failure and \
-       reboot_host_on_xen_livepatch_failure in \
-       host.pending_guidances_recommended"
-  ; version= (fun _ -> true)
+      "Upgrade pending update gudiances"
+      (* TODO: update below schema version to which the feature branch got merged with *)
```
Comment (minglumlu) [inline ocaml/xapi/xapi_db_upgrade.ml:872]: The schema version is being bumped up in this PR. This `TODO` can be removed.

**Review by minglumlu — APPROVED** [review]

**Comment by gangj** [comment]
> The schema version is being bumped up in this PR. This `TODO` can be removed.

I think we can keep this comment now and remove it when the feature branch got merged into master.

File: ocaml/idl/datamodel_common.ml [hunk 8]
```diff
@@ -10,7 +10,9 @@ open Datamodel_roles
               to leave a gap for potential hotfixes needing to increment the schema version.*)
 let schema_major_vsn = 5
 
-let schema_minor_vsn = 774
+(* This comment is a reminder to update the version in upgrade_update_guidance if schema_minor_vsn
+   changed when the update guidance improvement feature branch got merged into master branch *)
+let schema_minor_vsn = 775
```
Comment (robhoes) [inline ocaml/idl/datamodel_common.ml:15]: Please add a `let nile_release_schema_major_vsn` and `let nile_release_schema_minor_vsn` at the bottom of the list below, and refer to it from the upgrade rule.

**Comment by gangj** [comment]
> Please add a `let nile_release_schema_major_vsn` and `let nile_release_schema_minor_vsn` at the bottom of the list below, and refer to it from the upgrade rule.

OK, added

**Review by robhoes — APPROVED** [review]
